src/control_stream.ts

```
@@ -17,47 +17,73 @@ export class ControlStream {
   }
 
   async handshake() {
-    const writer = this.writer.getWriter();
-    await writer.write(
-      new ClientSetupEncoder({
+    console.log("handshaking...");
+    console.log(
+      "current supported draft: ",
+      CURRENT_SUPPORTED_DRAFT.toString(16),
+    );
+
+    try {
+      const m = new ClientSetupEncoder({
         type: MessageType.ClientSetup,
         versions: [CURRENT_SUPPORTED_DRAFT],
         parameters: [
           new ParameterEncoder({ type: 0, value: new Uint8Array([0x2]) }),
         ],
-      }),
-    );
-    writer.releaseLock();
+      });
+      await this.send(m);
+      console.log("ClientSetup msg sent");
+    } catch (error) {
+      console.log("failed to send ClientSetup msg: ", error);
+    }
 
     const reader = this.reader.getReader();
-    const { value, done } = await reader.read();
-    if (done) {
-      throw new Error("control stream closed");
-    }
-    if (value.type != MessageType.ServerSetup) {
-      throw new Error("invalid first message on control stream");
+    try {
+      const { value, done } = await reader.read();
+      if (done) {
+        throw new Error("control stream closed");
+      }
+      if (value.type != MessageType.ServerSetup) {
+        throw new Error("invalid first message on control stream");
+      }
+      console.log("handshake msg received: ", value);
+    } catch (error) {
+      console.log("failed to read handshake msg: ", error);
```

This will catch the invalid first message error we throw in line 47, which is supposed to inform the caller of the error, so I think we also need to re-throw here.

---

Since the [caller (`await controlStream.handshake();`)](https://github.com/mengelbart/moqjs/blob/writeObjUniStream/src/session.ts#L81) of `handshake()` is not catching errors (not surrounded by a try-catch block), we can just log out the error and return early, so errors are handled inside `handshake()` itself.

---

But that hides the error from the application that is calling `connect`. If an error can't be handled inside the library (e.g., by retrying), we need to surface it in the application so it can decide what to do next.

---

Oh, right. I am catching error from `connect()` too lmao, gotta throw the error to upstream. Thanks for clarifying :D